### PR TITLE
CDRIVER-4026 Specify 5.0 cursor behaviour

### PR DESCRIPTION
**Pull request by alcaeus**
CDRIVER-4026

This reverts changes from #793 and updates to the newest spec tests.

### CODE REVIEW

File: src/libmongoc/src/mongoc/mongoc-cursor.c [hunk 1]
```diff
@@ -1709,7 +1709,7 @@ _mongoc_cursor_prepare_getmore_command (mongoc_cursor_t *cursor,
       bson_append_int64 (command,
                          MONGOC_CURSOR_BATCH_SIZE,
                          MONGOC_CURSOR_BATCH_SIZE_LEN,
-                         batch_size);
+                         abs (_mongoc_n_return (cursor)));
```
Comment (jmikola) [inline src/libmongoc/src/mongoc/mongoc-cursor.c:1712]: Just so I understand, this change is prompted by the following from [the corresponding spec change](https://github.com/mongodb/specifications/commit/9da20b4b28f99037215ddfd78f5b0df04897a0be), correct?

> When setting the **batchSize** on the **find** and **getMore** commands the value of **batchSize** **MUST** be based on the cursor limit calculations specified in the `CRUD`_ specification.

Is there a particular reason `abs()` is needed here? Looking in existing uses of `_mongoc_n_return`, it looks like it was only used in `mongoc-cursor-legacy.c` for legacy OP_QUERY and OP_GET_MORE operations. In that case, I suppose a negative value was relevant (at least for [OP_QUERY](https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/#op_query)) in that it'd request as many documents and immediately close the cursor.

**Review by jmikola — APPROVED** [review]
LGTM. I think I probably answered my own question about `_mongoc_n_return`.

As for the behavioral impact on libmongoc, I assume this will just result in 5.0+ cursors needing to be explicitly killed since full iteration will still leave us with a cursor ID? I imagine that's a general performance regression from drivers due to the server-side behavioral change that you already talked through in https://github.com/mongodb/specifications/pull/1002.

File: src/libmongoc/src/mongoc/mongoc-cursor.c (hunk 1 above)
Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-cursor.c:1712]: I'm not sure about the exact motivations for using `abs()`, this merely restores functionality to what it was before I worked around the 5.0 issue. We'll be revisiting this logic in DRIVERS-1448, which will be done once we drop support for MongoDB < 3.6, which allows us to remove all `OP_QUERY` and `OP_GET_MORE` logic, including having the driver track the cursor limit.

File: src/libmongoc/tests/test-mongoc-cursor.c [hunk 2]
```diff
@@ -1907,7 +1907,7 @@ _test_cursor_n_return_find_cmd (mongoc_cursor_t *cursor,
             &getmore_cmd, "batchSize", tmp_bson ("{'$exists': false}"));
       }
 
-      assert_match_bson (request_get_doc (request, 0), &getmore_cmd, true);
+      ASSERT (match_bson (request_get_doc (request, 0), &getmore_cmd, true));
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/libmongoc/tests/test-mongoc-cursor.c]: Nope, this is merely due to me reverting the changes from #793. I've restored the use of `assert_match_bson`.

File: src/libmongoc/tests/json/command_monitoring/find.json [hunk 3]
```diff
@@ -356,7 +356,7 @@
               },
               "collection": "test",
               "batchSize": {
-                "$numberLong": "3"
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/libmongoc/tests/json/command_monitoring/find.json:359]: That is correct. We modified the tests to account for us not reducing the batchSize for the last batch, which is reverted now. I've also copied the newest version of both spec tests from the specs repo, so these are now synced to mongodb/specifications@9da20b4b. I've noted this in the original commit to preserve the information when we merge the pull request.

**Pull request merged**